pact-jvm-consumer-kotlin/README.adoc

```diff
@@ -0,0 +1,156 @@
+= kPact - Kotlin idiomatic DSL for Pact
+:p: icon:check[role="green"]
+:n: icon:times[role="red"]
+
+== Goals
+* Easy readable
+* Easy writable
+* Compile time safety (e.g. force mandatory data to be set)
+* short and concise (no boilerplate)
+* IDE support (e.g. auto-completion)
+* Things belonging to each other are in the same block
+* Things on the same logical level are on the same indentation level (auto-indentation of IDE)
+* Usable also from https://github.com/DiUS/pact-jvm/tree/master/pact-jvm-consumer-junit5[pact-jvm-consumer-junit5]
+  (from `PactDslWithProvider` to `RequestResponsePact`)
+* Support both:
+  1. prose like infix syntax (e.g. `"consumer" and "provider" havePact`)
+  2. fluent interface syntax (e.g. `consumer("consumer").hasPactWith("provider")`)
+
+== Examples
+
+[source,java,indent=0,role="primary"]
+.Infix DSL
+----
+include::src/test/kotlin/au/com/dius/pact/consumer/kotlin/dsl/examples/InfixDslExample.kt[lines=11..47]
+----
+
+[source,java,indent=0,role="secondary"]
+.Fluent DSL
+----
+include::src/test/kotlin/au/com/dius/pact/consumer/kotlin/dsl/examples/FluentDslExample.kt[lines=11..47]
+
+----
+
+[source,java,indent=0,role="secondary"]
+.ConsumerPactBuilder
+----
+include::src/test/kotlin/au/com/dius/pact/consumer/kotlin/dsl/examples/ConsumerPactBuilderExample.kt[lines=7..43]
+----
+
+== Usage together with https://github.com/DiUS/pact-jvm/tree/master/pact-jvm-consumer-junit5[pact-jvm-consumer-junit5]
+
+[source,java,indent=0,role="primary"]
+.kPact
+----
+include::src/test/kotlin/au/com/dius/pact/consumer/kotlin/dsl/examples/Junit5Example.kt[lines=14..26]
+----
+
+[source,java,indent=0,role="secondary"]
+.ConsumerPactBuilder
+----
+include::src/test/kotlin/au/com/dius/pact/consumer/kotlin/dsl/examples/Junit5Example.kt[lines=30..41]
+----
+
+== Open Design Decisions
+
+=== kPact block VS enforcing single Consumer/Provider pair
+
+==== Examples
+
+[source,java,indent=0,role="primary"]
+.kpact block example
+----
+kPact {
+    consumer withName "consumer" andProvider "provider" havePact {
+        // ...
+    }
+}
+----
+
+[source,java,indent=0,role="secondary"]
+.enforcing single Consumer/Provider pair example
+----
+KPact between "consumer" andProvider "provider" isDefinedBy {
+    // ...
+}
+----
+
+==== (Dis-)Advantages
+
+===== kpact block
+
+{n}
+  **compile time safety** / unexpected and inconsistent behavior: +
+  Multiple consumer/provider blocks are not prevented by compiler but
+  only last one would be used:
+[source,java,indent=0]
+----
+      kPact {
+          consumer withName "consumer" andProvider "provider" havePact { // <1>
+              // ...
+          }
+
+          consumer withName "consumer" andProvider "provider" havePact {
```

Comment:
I'm not sure I understand your concern here. Couldn't this second call raise an exception, due to a conflict with an existing pact?

Reply:
The problem is the missing compile time safety. The exception is thrown during runtime, so the compiler and the IDE cannot help to prevent this error. With the other approach this error is not possible at all.